File: leet_code/js/sortByBits.js
// https://leetcode.com/problems/sort-integers-by-the-number-of-1-bits/
// time: 76ms, memory: 35.9MB

/**
 * @param {number[]} arr
 * @return {number[]}
 */
var sortByBits = function(arr) {
  function countBits(n) {
    let count = 0;
    while (n > 0) {
      count += n & 1;
      n >>= 1;
    }
    return count;
  }

  return arr.sort((a, b) => countBits(a) - countBits(b) || a - b);
};

File: leet_code/c++/sort_by_bits.cpp
// https://leetcode.com/problems/sort-integers-by-the-number-of-1-bits/
// time: 16ms, memory: 8.8MB
#include <bitset>
#include <string>
#include <vector>

using namespace std;

int countBits(int n) {
  bitset<16> b(n);
  return b.count();
}

bool comparator(int a, int b) {
  int an = countBits(a);
  int bn = countBits(b);
  return an == bn ? a < b : an < bn;
}

class Solution {
 public:
  vector<int> sortByBits(vector<int>& arr) {
    sort(arr.begin(), arr.end(), comparator);

    return arr;
  }
};